From 6e8c32fe7c00bceee13fdce553a31501c9208866 Mon Sep 17 00:00:00 2001
From: Sebastian Wick <sebastian.wick@redhat.com>
Date: Thu, 3 Feb 2022 17:01:41 +0100
Subject: [PATCH] window-group: Disable culling for ClutterOffscreenEffects

Culling is currently disabled for ClutterOffscreenEffects only if the
MetaWindowGroup is painted as part of a ClutterClone. This results in
glitches if a ClutterOffscreenEffect is attached to the MetaWindowGroup
or any parents. Set the clip region to the entire stage when the context
FB and the stage FB are different.
---
 src/compositor/meta-window-group.c | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/src/compositor/meta-window-group.c b/src/compositor/meta-window-group.c
index 16ff2db097..367d711b11 100644
--- a/src/compositor/meta-window-group.c
+++ b/src/compositor/meta-window-group.c
@@ -65,6 +65,8 @@ meta_window_group_paint (ClutterActor        *actor,
   cairo_rectangle_int_t visible_rect;
   int paint_x_origin, paint_y_origin;
   int screen_width, screen_height;
+  CoglFramebuffer *fb = clutter_paint_context_get_framebuffer (paint_context);
+  ClutterStageView *view = clutter_paint_context_get_stage_view (paint_context);
 
   redraw_clip = clutter_paint_context_get_redraw_clip (paint_context);
   if (!redraw_clip)
@@ -90,14 +92,9 @@ meta_window_group_paint (ClutterActor
    */
   if (clutter_actor_is_in_clone_paint (actor))
     {
-      CoglFramebuffer *fb;
-      ClutterStageView *view;
       MetaTransforms trans;
 
-      fb = clutter_paint_context_get_framebuffer (paint_context);
-      view = clutter_paint_context_get_stage_view (paint_context);
       if (!view ||
-          fb != clutter_stage_view_get_framebuffer (view) ||
           !meta_actor_painting_untransformed (fb,
                                               screen_width,
                                               screen_height,
@@ -127,7 +123,15 @@ meta_window_group_paint (ClutterActor        *actor,
    * multihead setup with mismatched monitor sizes, we could intersect this
    * with an accurate union of the monitors to avoid painting shadows that are
    * visible only in the holes. */
-  clip_region = cairo_region_copy (redraw_clip);
+  if (view && fb == clutter_stage_view_get_framebuffer (view))
+    {
+      clip_region = cairo_region_copy (redraw_clip);
+      cairo_region_translate (clip_region, -paint_x_origin, -paint_y_origin);
+    }
+  else
+    {
+      clip_region = cairo_region_copy (unobscured_region);
+    }
 
   cairo_region_translate (clip_region, -paint_x_origin, -paint_y_origin);
 
-- 
GitLab

